runtime/sema/check_assignment.go

```diff
@@ -344,7 +347,12 @@ func (checker *Checker) visitIndexExpressionAssignment(
 	indexExpression *ast.IndexExpression,
 ) (elementType Type) {
 
-	elementType = checker.visitIndexExpression(indexExpression, true)
+	// in an statement like `ref.foo[i] = x`, the entire statement itself
+	// is an assignment, but the evaluation of the index expression itself (i.e. `ref.foo`)
+	// is not, so we temporarily clear the `inAssignment` status here before restoring it later.
+	elementType = checker.withAssignment(false, func() Type {
+		return checker.visitIndexExpression(indexExpression, true)
+	})
```

Comment:
For the member expressions, the same check is done in the general method `visitMember` (not in the assignment specific method `visitMemberExpressionAssignment`). That makes sense, since it will handle all cases.

Similarly here, wouldn't it be better to do this change in the `visitIndexExpression` generic method, so it'l handle for all cases, similar to member-expressions?
Also, I think here what needs wrapping is the visiting of `indexExpression.TargetExpression`, not the `indexExpression` itself, right?